File: ROS/src/simple_ros/simple_ros/move_turtle.py
import rclpy
from geometry_msgs.msg import Twist
from rclpy.node import Node
from rclpy.qos import (
    QoSDurabilityPolicy,
    QoSHistoryPolicy,
    QoSProfile,
    QoSReliabilityPolicy,
)
from turtlesim.msg import Color, Pose


class Move_turtle(Node):
    def __init__(self):
        super().__init__("hello_pub")
        self.qos_profile = QoSProfile(history=QoSHistoryPolicy.KEEP_LAST,
                                      reliability=QoSReliabilityPolicy.RELIABLE,
                                      durability=QoSDurabilityPolicy.VOLATILE,
                                      depth=10)
        self.create_timer(0.1, self.twist_pub)
        self.create_timer(1/60, self.update)
        self.pub = self.create_publisher(Twist, "turtle1/cmd_vel", self.qos_profile)
        self.create_subscription(Pose, "turtle1/pose", self.pose_callback, self.qos_profile)
        self.create_subscription(Color, "turtle1/color_sensor", self.color_callback, self.qos_profile)
        self.twist = Twist()
        self.pose = Pose()
        self.color = Color()
        self.phase = 0

    def twist_pub(self):
        self.pub.publish(self.twist)

    def pose_callback(self, msg: Pose):
        self.pose = msg

    def color_callback(self, msg: Color):
        self.color = msg
    
    def update(self):
        """ self.twist, self.pose, self.color 을 이용한 알고리즘"""
        # self.twist.linear.x += 0.001
        # self.twist.angular.z = 1.0
        if self.phase == 0:
            if self.pose.theta < 0:
                self.twist.linear.x = 0.0
                self.twist.angular.z = 1.0
            elif self.pose.x < 8:
                self.twist.linear.x = 0.6
                self.twist.angular.z = 0.0
            else:
                self.phase = 1
        elif self.phase == 1:
            if self.pose.theta < 3.141592/2:
                self.twist.linear.x = 0.0
                self.twist.angular.z = 1.0
            elif self.pose.y < 8:
                self.twist.linear.x = 0.6
                self.twist.angular.z = 0.0
            else:
                self.phase = 2
        elif self.phase == 2:
            if self.pose.theta < 3.0:
                self.twist.linear.x = 0.0
                self.twist.angular.z = 1.0
            elif self.pose.x > 2:
                self.twist.linear.x = 0.6
                self.twist.angular.z = 0.0
            else:
                self.phase = 3
                self.get_logger().info("phase 3")
        elif self.phase == 3:
            if  not -3.341592/2 < self.pose.theta < -3.141592/2:
                self.twist.linear.x = 0.0
                self.twist.angular.z = 1.0
            elif self.pose.y > 2:
                self.twist.linear.x = 0.6
                self.twist.angular.z = 0.0
            else:
                self.phase = 0

def main():
    rclpy.init()
    node = Move_turtle()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.destroy_node()

if __name__ == "__main__":
    main()

File: ROS/src/simple_ros_cpp/src/move_turtle.cpp
#include "geometry_msgs/msg/twist.hpp"
#include "rclcpp/rclcpp.hpp"
#include "turtlesim/msg/color.hpp"
#include "turtlesim/msg/pose.hpp"
#include <chrono>
#include <iostream>

using namespace std;
using namespace std::chrono_literals;

class MoveTurtle : public rclcpp::Node
{
public:
    MoveTurtle()
        : Node("hello_sub")
    {
        auto qos_profile = rclcpp::QoS(rclcpp::KeepLast(10))
                               .reliability(RMW_QOS_POLICY_RELIABILITY_RELIABLE)
                               .durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);
        _pose_sub = create_subscription<turtlesim::msg::Pose>(
            "turtle1/pose",
            qos_profile,
            std::bind(&MoveTurtle::pose_sub_callback, this, std::placeholders::_1));
        _color_sub = create_subscription<turtlesim::msg::Color>(
            "turtle1/color",
            qos_profile,
            std::bind(&MoveTurtle::color_sub_callback, this, std::placeholders::_1));
        _pub = create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", qos_profile);
        _twist_pub_timer = create_wall_timer(100ms, std::bind(&MoveTurtle::twist_pub, this));
        // _update_timer = create_wall_timer(std::chrono::milliseconds(17),...)
        _update_timer = create_wall_timer(17ms, std::bind(&MoveTurtle::update, this));
    }

private:
    int _count;
    turtlesim::msg::Pose _pose;
    turtlesim::msg::Color _color;
    geometry_msgs::msg::Twist _twist;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr _pose_sub;
    rclcpp::Subscription<turtlesim::msg::Color>::SharedPtr _color_sub;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr _pub;
    rclcpp::TimerBase::SharedPtr _twist_pub_timer;
    rclcpp::TimerBase::SharedPtr _update_timer;
    void pose_sub_callback(const turtlesim::msg::Pose::SharedPtr msg)
    {
        _pose = *msg;
    }
    void color_sub_callback(const turtlesim::msg::Color::SharedPtr msg)
    {
        _color = *msg;
    }
    void twist_pub()
    {
        _pub->publish(_twist);
    }
    void update()
    {
        _twist.linear.x += 0.006;
        _twist.angular.z = 1.0;
    }
};

int main()
{
    rclcpp::init(0, nullptr);
    auto node = std::make_shared<MoveTurtle>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
